SignalServiceKit/src/Network/OutageDetection.swift

```diff
@@ -0,0 +1,122 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import os
+
+@objc
+public class OutageDetection: NSObject {
+    @objc(sharedManager)
+    public static let shared = OutageDetection()
+
+    @objc public static let outageStateDidChange = Notification.Name("OutageStateDidChange")
+
+    // These properties should only be accessed on the main thread.
+    @objc
+    public var hasOutage = false {
+        didSet {
+            SwiftAssertIsOnMainThread(#function)
+
+            NotificationCenter.default.postNotificationNameAsync(OutageDetection.outageStateDidChange, object: nil)
+        }
+    }
+    private var mayHaveOutage = false {
+        didSet {
+            SwiftAssertIsOnMainThread(#function)
+
+            ensureCheckTimer()
+        }
+    }
+
+    // We only show the outage warning when we're certain there's an outage.
+    // DNS lookup failures, etc. are not considered an outage.
+    private func checkForOutageSync() -> Bool {
+        let host = CFHostCreateWithName(nil, "uptime.signal.org" as CFString).takeRetainedValue()
+        CFHostStartInfoResolution(host, .addresses, nil)
+        var success: DarwinBoolean = false
+        guard let addresses = CFHostGetAddressing(host, &success)?.takeUnretainedValue() as NSArray? else {
```

Comment:
don't we want to takeRetainedValue here?

Reply:
I could be wrong, but here's my take: you only want to use `takeRetainedValue` when the object in question's ownership is transferred to you, e.g. by a `copy` or `create` method or function.  `CFHostGetAddressing` is a `get` function and doesn't transfer ownership, so it doesn't increment the retain count, so it wouldn't be safe for us to decrement the retain count, so we shouldn't `takeRetainedValue`.